### PR TITLE
Command Migration: JSON. ARRAPPEND, ARRPOP, ARRLEN

### LINKED ISSUE #1026: Command Migration: ('JSON.ARRAPPEND', 'JSON.ARRLEN', 'JSON.ARRPOP')
This issue tracks the migration of the mentioned commands - ('JSON.ARRAPPEND', 'JSON.ARRLEN', 'JSON.ARRPOP') to make them compatible across the three protocols supported by the Dice database: RESP, HTTP, and WebSocket.
DiceDB now supports more than one protocols (Resp/http/websocket). We don't want eval operation of each command to be strictly bound with any of the protocols. Currently eval function return statements of each command is bounded to the RESP protocol. 

*The migration is required because:*

- The current implementation is specific to the RESP protocol
- Evaluation functions of each command must be refactored to ensure protocol-agnostic behavior.
- Evaluation functions should return raw responses instead of protocol specific responses.

The goal is to make the command logic protocol-independent, allowing all three protocols to call the same core functionality seamlessly.

### Requirements

- Refactor the evaluation function of `[Command Name]` to be generic.
- Ensure the evaluation function does not include protocol-specific logic.

### Migration Steps

1. **Analyze Current Implementation**
   - Review the current codebase to understand how the command logic is implemented.

2. **Refactor Return Logic**
   - Create a new function with the function definition as `evalXXX(args []string, store *dstore.Store) *EvalResponse` under the file `/internal/eval/store_eval.go`
   - Analyse return statements of the eval and modify them to send raw types without encoding
   - Use errors in the return statements from `/errors/migrated_errors.go` file
   - Use perdefined responses from `/internal/clientio/resp.go` file as RespType variables

3. **Command/Worker specific Changes**
   - Make `IsMigrated` flag to true in the commands information under `/internal/eval/commands.go` file.
   - Use the newly written eval function against the `NewEval` parameter in the same command structure.
   - Delete the old eval function from the `/internal/eval/eval.go` file.
   - Add the migrated command to the `/internal/worker/CommandsMeta` map and make it's type as `SingleShard`

4. **Update Unit Tests**
   - Refactor existing unit tests to accommodate the new implementation.
   - Add new unit tests if necessary to cover all possible cases.

5. **Integration Tests**
   - Run all integration tests to ensure successful migration.
   - Ensure that each protocol (RESP, HTTP, WebSocket) works correctly after migration.

### Checklist

- [ ] Migrated the evalXXX function with the latest definition
- [ ] Update or add unit tests for the new implementation.
- [ ] All unit tests pass successfully.
- [ ] Ensure all integration tests pass successfully.

### Additional Notes

- Describe any edge cases that need to be handled in the generic function.
- Mention any protocol-specific optimizations required in wrappers.
  
If there are any questions or concerns about this migration, please mention them here.

### Related Issues/PRs

- Sample implementation for the Get, Set, GetSet and SetEx can be found in [this Pull Request.](https://github.com/DiceDB/dice/pull/766)

### PR DESCRIPTION
**Pull request by srivastava-yash**
This PR migrates JSON.ARRAPPEND, JSON.ARRPOP, JSON.ARRLEN to make it protocol agnostic. Closes #1026 

Migration Checklist
- [x]  Migrated the evalXXX function with the latest definition
- [x]  Update or add unit tests for the new implementation.
- [x]  All unit tests pass successfully.
- [x]  Ensure all integration tests pass successfully.
- [x] Ensure integration tests are added for the migrated command on multi-threaded resp server.
- [x] Move Integration tests for the respective commands under the RESP integration tests directory from Async directory
- [x] Please validate that the documentation for the respective commands is up to date. If not then consider adding them.

### CODE REVIEW

**Comment by AshwinKul28** [comment]
@srivastava-yash Thanks for the PR. 
Can you please check the failed test cases?

**Comment by srivastava-yash** [comment]
> @srivastava-yash Thanks for the PR. Can you please check the failed test cases?

yes @AshwinKul28, working on it.

**Comment by srivastava-yash** [comment]
The integration tests for the commands ARRAPPEN and ARRPOP seem to be failing because of the same reason, when ever the response is to be nil its getting a different string for nil. As seen in the attached screenshot the expected is (nil) but its getting <nil>. Changing expected to <nil> does not fix the test case either. Still finding reason for this.
image

File: internal/eval/store_eval.go [hunk 1]
```diff
@@ -316,3 +318,367 @@ func evalSETEX(args []string, store *dstore.Store) *EvalResponse {
...
+
+	// process value at each path
+	popArr := make([]any, 0, len(results))
+	for _, result := range results {
+		arr, ok := result.([]any)
+		// if value can not be converted to array, it is of another type
+		// returns nil in this case similar to redis
+		// also, return nil if array is empty
+		if !ok || len(arr) == 0 {
+			popArr = append(popArr, nil)
```
Comment (AshwinKul28) [inline internal/eval/store_eval.go:657]: Looks like here a statement is missed
`popElem := clientio.NIL`

**Review by AshwinKul28 — CHANGES_REQUESTED** [review]
Thanks a lot @srivastava-yash  for the changes. Left few comments.

File: internal/eval/store_eval.go (hunk 1 above)
Comment (srivastava-yash) [reply internal/eval/store_eval.go:657]: clientio.NIL will output 0 right, are we not sending nil as response now?

Comment (AshwinKul28) [reply internal/eval/store_eval.go:657]: When you send clientio.NIL from eval, it works as an Enum and converts output response to `[]byte("$-1\r\n")` which resp decodes as `(nil)`.

**Comment by soumya-codes** [comment]
@srivastava-yash lets ensure that the corresponding integration tests are migrated as well.

**Comment by soumya-codes** [comment]
Just FYI, we need to add the tests for the migrated commands under integration_tests/commands/resp

I have added a new item to checklist in the PR description. Please tick the same once the integration tests are added.
Please let me or Ashwin know if you have any doubts/concern.

File: internal/eval/store_eval.go [hunk 2]
```diff
@@ -316,3 +318,367 @@ func evalSETEX(args []string, store *dstore.Store) *EvalResponse {
...
+	}
+
+	key := args[0]
+	path := args[1]
+	values := args[2:]
+
+	obj := store.Get(key)
+	if obj == nil {
+		return &EvalResponse{
+			Result: nil,
```
Comment (AshwinKul28) [inline internal/eval/store_eval.go:343]: It looks like here you need to return clientio.NIL, this should solve the HTTP issue.

File: internal/eval/store_eval.go [hunk 3]
```diff
@@ -316,3 +318,367 @@ func evalSETEX(args []string, store *dstore.Store) *EvalResponse {
...
+	}
+	key := args[0]
+
+	// Retrieve the object from the database
+	obj := store.Get(key)
+
+	// If the object is not present in the store or if its nil, then we should simply return nil.
+	if obj == nil {
+		return &EvalResponse{
+			Result: nil,
```
Comment (AshwinKul28) [inline internal/eval/store_eval.go:445]: It looks like here you need to return clientio.NIL, this should solve the HTTP issue.

File: internal/eval/store_eval.go [hunk 4]
```diff
@@ -316,3 +318,367 @@ func evalSETEX(args []string, store *dstore.Store) *EvalResponse {
...
+	var index string
+	if len(args) >= 3 {
+		index = args[2]
+	}
+
+	// Retrieve the object from the database
+	obj := store.Get(key)
+	if obj == nil {
+		return &EvalResponse{
+			Result: nil,
```
Comment (AshwinKul28) [inline internal/eval/store_eval.go:588]: It looks like here you need to return clientio.NIL, this should solve the HTTP issue.

File: internal/eval/store_eval.go (hunk 1 above)
Comment (AshwinKul28) [reply internal/eval/store_eval.go:657]: Okay I got the issue, In this case eval is returning []interface and in that 1st element is of type clientio.NIL. hence the following HTTP server write code is unable to detect that. 

```
	respArr := []string{
		"(nil)",  // Represents a RESP Nil Bulk String, which indicates a null value.
		"OK",     // Represents a RESP Simple String with value "OK".
		"QUEUED", // Represents a Simple String indicating that a command has been queued.
		"0",      // Represents a RESP Integer with value 0.
		"1",      // Represents a RESP Integer with value 1.
		"-1",     // Represents a RESP Integer with value -1.
		"-2",     // Represents a RESP Integer with value -2.
		"*0",     // Represents an empty RESP Array.
	}


	if val, ok := responseValue.(clientio.RespType); ok {
		responseValue = respArr[val]
	}
```

@pratikpandey21 sir your help is much appreciated. we can iterate through array and write one case for that, but you would be the best judge here.

Comment (AshwinKul28) [reply internal/eval/store_eval.go:657]: btw @srivastava-yash instead of adding nil, please consider appending clientio.NIL to this slice.

Comment (srivastava-yash) [reply internal/eval/store_eval.go:657]: > btw @srivastava-yash instead of adding nil, please consider appending clientio.NIL to this slice.

yes @AshwinKul28 , I have already made that change. Haven't pushed it yet. I will do that

Comment (AshwinKul28) [reply internal/eval/store_eval.go]: @lucifercr07 Please have a look, was talking about this case.

**Comment by lucifercr07** [comment]
@srivastava-yash please resolve the conflicts once.

**Comment by srivastava-yash** [comment]
> @srivastava-yash please resolve the conflicts once.

working on it @lucifercr07

**Comment by AshwinKul28** [comment]
@srivastava-yash, please consider working on other items on the checklist. I know it's a bit more effort, but since you are the best person to work on these complimentary tasks related to these commands. Thanks

**Comment by lucifercr07** [comment]
@srivastava-yash please resolve the conflicts.

**Comment by srivastava-yash** [comment]
> @srivastava-yash please resolve the conflicts.

Done @lucifercr07

**Comment by lucifercr07** [comment]
@srivastava-yash HTTP issue has been fixed as part of #1150, is there any other blocker for this?

**Comment by srivastava-yash** [comment]
> @srivastava-yash HTTP issue has been fixed as part of #1150, is there any other blocker for this?

Thanks @lucifercr07, I missed the PR. I will resolve conflict and fix integration tests

**Comment by srivastava-yash** [comment]
@lucifercr07 @AshwinKul28 PR is up for review again. Completed all the tasks and resolved conflicts. Thank you!

**Comment by AshwinKul28** [comment]
@srivastava-yash PR looks fantastic. Thanks a lot for your commendable efforts. 

I see the documentation for `ARRLEN` is not available. I know it's little back and forth but we want to make this PR production ready to get it merged, so we never needs to visit this again. Please have a look at this[ sample documentation](https://github.com/DiceDB/dice/blob/master/docs/sample_command_docs.md). 

Please make changes as follows:
1. Feel free to remove numbering for examples
2. Add description for each example
3. Add additional notes and best practices if required. 

Once these cosmetic changes are in, we are good to merge. Use LLMs to generate docs faster. 

Thanks a lot again.

**Comment by srivastava-yash** [comment]
@AshwinKul28 the documentation for ARRLEN is already there and did not require any updates. ARRPOP was missing which I added with this PR.

**Comment by apoorvyadav1111** [comment]
Hi @srivastava-yash , thank you for adding documentation for `ARRPOP`. It looks good but needs some minor changes in terms of structure. Please use this template for docs: docs/command_docs_template.md. This will ensure we have consistency in the webpages.

Thanks

**Comment by AshwinKul28** [comment]
@srivastava-yash Can you please take care of the conflicts and the one that @apoorvyadav1111 mentioned? Then we are good to go! Thanks

**Comment by srivastava-yash** [comment]
@AshwinKul28 @apoorvyadav1111 yes sure working on it. Sorry the last couple of days were a little busy and I couldnt work on it. Doing this today

**Comment by srivastava-yash** [comment]
@AshwinKul28 Done!

File: docs/src/content/docs/commands/JSON.ARRPOP.md [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+
+When the `JSON.ARRPOP` command is executed, the specified element is popped from the array located at the given index at the given path within the JSON document stored under the specified key. If the path does not exist or does not point to an array, an error will be raised.
+
+## Error Handling
+
+- `(error) ERR key does not exist`: Raised if the specified key does not exist in the DiceDB database.
+- `(error) ERR path %s does not exist`: Raised if the specified path, s, does not exist within the JSON document.
+- `(error) ERR path is not an array`: Raised if the specified path does not point to a JSON array.
+- `(error) WRONGTYPE Operation against a key holding the wrong kind of value`: Raised if any of the provided JSON values are not valid JSON.
+
```
Comment (apoorvyadav1111) [inline docs/src/content/docs/commands/JSON.ARRPOP.md:36]: Please follow below bullet structure for errors in both files
1. `Wrong type of value or key`:

   - Error Message: `(error) WRONGTYPE Operation against a key holding the wrong kind of value`
   - Occurs when attempting to use the command on a key that contains a non-string value.

File: docs/src/content/docs/commands/JSON.ARRAPPEND.md [hunk 6]
```diff
@@ -28,9 +28,9 @@ When the `JSON.ARRAPPEND` command is executed, the specified JSON values are app
 ## Error Handling
 
 - `(error) ERR key does not exist`: Raised if the specified key does not exist in the DiceDB database.
-- `(error) ERR path does not exist`: Raised if the specified path does not exist within the JSON document.
+- `(error) ERR path %s does not exist`: Raised if the specified path, s, does not exist within the JSON document.
 - `(error) ERR path is not an array`: Raised if the specified path does not point to a JSON array.
-- `(error) ERR invalid JSON`: Raised if any of the provided JSON values are not valid JSON.
+- `(error) WRONGTYPE Operation against a key holding the wrong kind of value`: Raised if any of the provided JSON values are not valid JSON.
```
Comment (apoorvyadav1111) [inline docs/src/content/docs/commands/JSON.ARRAPPEND.md:34]: same as ARRPOP

**Review by apoorvyadav1111 — CHANGES_REQUESTED** [review]
As its already reviewed from @AshwinKul28 , I just reviewed for docs and I would like to request changes in the heading and bullet structures on all three files. Additionally, lets use bash style code block instead of plaintext. Lastly, please confirm it the existing ARRLEN documentation follownthese standards. 

Again, we would like to thank you for your efforts in this monumental change. You can always reach out to us on discord and here if you are blocked or don't have bandwidth, this allows us to adjust our goals and keep your issue assigned to you.

Thanks,
Apoorv

**Comment by srivastava-yash** [comment]
Thanks for the review comments. Done @apoorvyadav1111 !

**Review by apoorvyadav1111 — APPROVED** [review]
Hi @srivastava-yash, thanks for making the changes, we are close to merging this one. I am unable to find websocket tests for these commands. Can you please confirm if they are sufficient if available or add them? Additionally, this next request is a nit but will definitely help us to ensure consistency. Could you please change the below heading to Example Usage in arrlen docs?
image

**Comment by srivastava-yash** [comment]
@AshwinKul28 @lucifercr07 getting the same error for websocket as what we were getting for the case of HTTP. Getting 0 instead of nil.
internal/server/websocketserver.go -> line 266
image
I think its because of the same code piece that was earlier there in HTTP as well

**Comment by AshwinKul28** [comment]
Thanks @srivastava-yash for pointing this. @lucifercr07 we may need to replicate the changes done by you in HTTP to websocket as well.

**Comment by lucifercr07** [comment]
@srivastava-yash yes it's the same issue, I've raised a PR for fixing similar issue https://github.com/DiceDB/dice/pull/1204. Need to add some more refactoring to that.

**Comment by AshwinKul28** [comment]
@lucifercr07 If we merge this PR, all other PRs with the latest code will fail, what are your thoughts?
